### PR TITLE
Fix appservices being unable to handle to_device messages for multiple users

### PR DESCRIPTION
**Pull request by Half-Shot**
This was a *really* subtle bug, but it caused a great deal of pain. In order to fetch to_device messages for an appservice, we must fetch all the devices an AS has control over. We were doing this filter, but mistakenly left in an extra clause that would filter for the *last* user_id in the set of users.

Why? Because Python is awful, and just above this statement there was a for loop that used the `user_id` variable to do some filter logic. This code worked because that variables was still in scope, and therefore the device query code looked like it worked...until you tried to use it with two user_ids.

Dropping this clause makes Synapse correctly calculate new messages for *all* users in the appservice namespace.

### CODE REVIEW

File: changelog.d/16251.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a longstanding bug where appservices using MSC2049 to receive to_device messages, would only get messages for one user.
```
Comment (clokep) [inline changelog.d/16251.bugfix:1]: MSC2049 doesn't seem to be a thing?

```suggestion
Fix a long-standing bug where appservices using MSC2049 to receive `to_device` messages, would only get messages for one user.
```

File: synapse/storage/databases/main/deviceinbox.py
Comment (clokep) [inline synapse/storage/databases/main/deviceinbox.py:1]: Can we add a test for this?

**Review by clokep — CHANGES_REQUESTED** [review]

File: synapse/storage/databases/main/deviceinbox.py
Comment (Half-Shot) [reply synapse/storage/databases/main/deviceinbox.py:1]: With great difficulty, there doesn't seem to be any deviceinbox tests... I can start that effort. I always seem to walk into these situations :laughing:

Comment (clokep) [reply synapse/storage/databases/main/deviceinbox.py:1]: #11215 seems to have some tests around this, can we expand `ApplicationServicesHandlerSendEventsTestCase` to cover the case you're fixing?

File: changelog.d/16251.bugfix (hunk 1 above)
Comment (AndrewFerr) [reply changelog.d/16251.bugfix:1]: The correct MSC is [2409](https://github.com/Sorunome/matrix-doc/blob/soru%2Bhs/appservice-edus/proposals/2409-appservice-edus.md).

File: synapse/storage/databases/main/deviceinbox.py
Comment (Half-Shot) [reply synapse/storage/databases/main/deviceinbox.py:1]: Well spotted, I have added one.

File: tests/handlers/test_appservice.py [hunk 2]
```diff
@@ -849,6 +861,125 @@ def test_application_services_receive_bursts_of_to_device(self) -> None:
...
+        for device_msg in [device_msg_1, device_msg_2, device_msg_3]:
+            self.assertEqual(device_msg["msg"]["type"], "m.room_key_request")
+            self.assertEqual(device_msg["msg"]["sender"], self.local_user)
+            self.assertEqual(
+                device_msg["msg"]["to_user_id"], device_msg["expected_user"]
+            )
+            self.assertEqual(
+                device_msg["msg"]["to_device_id"], device_msg["expected_device"]
+            )
+            self.assertEqual(device_msg["msg"]["content"], message_content)
```
Comment (clokep) [inline tests/handlers/test_appservice.py:981]: I think adding the in-between mapping is more confusing than just doing the asserts?

```suggestion
        for device_msg in to_device_messages:
            self.assertEqual(device_msg["type"], "m.room_key_request")
            self.assertEqual(device_msg["sender"], self.local_user)
            self.assertEqual(device_msg["content"], message_content)

        self.assertEqual(
            to_device_messages[0]["to_user_id"], self.exclusive_as_user
        )
        self.assertEqual(
            to_device_messages[0]["to_device_id"], self.exclusive_as_user_device_id,
        )

        self.assertEqual(
            to_device_messages[1]["to_user_id"], self.exclusive_as_user_2
        )
        self.assertEqual(
            to_device_messages[1]["to_device_id"], self.exclusive_as_user_2_device_id,
        )

        self.assertEqual(
            to_device_messages[2]["to_user_id"], self.exclusive_as_user_3
        )
        self.assertEqual(
            to_device_messages[2]["to_device_id"], self.exclusive_as_user_3_device_id,
        )
```

**Review by clokep — APPROVED** [review]
Looks good if CI is happy.

**Pull request merged**